File: macos/Tests/YOLOTermKitTests/Persistence/WorkspaceStoreTests.swift
import XCTest
@testable import YOLOTermKit

@MainActor
final class WorkspaceStoreTests: XCTestCase {
    
    var tempDirectory: URL!
    var store: WorkspaceStore!
    
    override func setUp() async throws {
        tempDirectory = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString, isDirectory: true)
        try FileManager.default.createDirectory(at: tempDirectory, withIntermediateDirectories: true)
        store = WorkspaceStore(baseDirectory: tempDirectory)
    }
    
    override func tearDown() async throws {
        try? FileManager.default.removeItem(at: tempDirectory)
    }
    
    // MARK: - Basic Save and Load
    
    func testSaveAndLoad() async throws {
        let workspace = WorkspaceStore.Workspace(
            tabs: [
                WorkspaceStore.Tab(
                    name: "Tab 1",
                    panes: [
                        WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/Users/test")
                    ],
                    layoutPreset: "single"
                )
            ],
            selectedTabIndex: 0
        )
        
        try await store.saveImmediate(workspace)
        let loaded = try store.load()
        
        XCTAssertEqual(loaded, workspace, "Loaded workspace should match saved workspace")
    }
    
    func testLoadNonExistent() throws {
        let loaded = try store.load()
        XCTAssertNil(loaded, "Loading non-existent workspace should return nil")
    }
    
    func testMultipleTabs() async throws {
        let workspace = WorkspaceStore.Workspace(
            tabs: [
                WorkspaceStore.Tab(
                    name: "Tab 1",
                    panes: [
                        WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/Users/test1")
                    ],
                    layoutPreset: "single"
                ),
                WorkspaceStore.Tab(
                    name: "Tab 2",
                    panes: [
                        WorkspaceStore.Pane(shell: "/bin/bash", cwd: "/Users/test2"),
                        WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/Users/test3")
                    ],
                    layoutPreset: "columns"
                )
            ],
            selectedTabIndex: 1
        )
        
        try await store.saveImmediate(workspace)
        let loaded = try store.load()
        
        XCTAssertEqual(loaded?.tabs.count, 2, "Should load 2 tabs")
        XCTAssertEqual(loaded?.selectedTabIndex, 1, "Should restore selected tab index")
        XCTAssertEqual(loaded?.tabs[1].panes.count, 2, "Tab 2 should have 2 panes")
    }
    
    // MARK: - Graceful Degradation (CRITICAL - TermGrid lesson)
    
    func testGracefulDegradation_InvalidJSON() throws {
        // TermGrid c65fef8 lesson: corrupt workspace should not lose all tabs
        
        try store.writeCorruptWorkspace(scenario: .invalidJSON)
        
        let loaded = try store.load()
        
        // Should return empty workspace, not crash or return nil
        XCTAssertNotNil(loaded, "Corrupt workspace should not return nil")
        XCTAssertEqual(loaded?.tabs.count, 0, "Corrupt workspace should return empty tabs")
    }
    
    func testGracefulDegradation_MissingTabs() throws {
        try store.writeCorruptWorkspace(scenario: .missingTabs)
        
        let loaded = try store.load()
        
        XCTAssertNotNil(loaded, "Workspace with missing tabs field should not return nil")
        XCTAssertEqual(loaded?.tabs.count, 0, "Workspace with missing tabs should return empty array")
    }
    
    func testGracefulDegradation_PartiallyCorruptTab() throws {
        // One valid tab, one corrupt tab - should salvage the valid one
        
        try store.writeCorruptWorkspace(scenario: .partiallyCorruptTab)
        
        let loaded = try store.load()
        
        XCTAssertNotNil(loaded, "Partially corrupt workspace should not return nil")
        XCTAssertEqual(loaded?.tabs.count, 1, "Should salvage the valid tab")
        XCTAssertEqual(loaded?.tabs.first?.name, "Tab 1", "Should preserve the valid tab's data")
    }
    
    func testNeverDestructiveOnPartialLoad() async throws {
        // CRITICAL TEST: This is the TermGrid c65fef8 bug test
        // If workspace is partially readable, we must preserve what we can
        
        // Create a workspace with multiple tabs
        let originalWorkspace = WorkspaceStore.Workspace(
            tabs: [
                WorkspaceStore.Tab(name: "Tab 1", panes: [
                    WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/Users/test1")
                ], layoutPreset: "single"),
                WorkspaceStore.Tab(name: "Tab 2", panes: [
                    WorkspaceStore.Pane(shell: "/bin/bash", cwd: "/Users/test2")
                ], layoutPreset: "single"),
                WorkspaceStore.Tab(name: "Tab 3", panes: [
                    WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/Users/test3")
                ], layoutPreset: "single")
            ],
            selectedTabIndex: 0
        )
        
        try await store.saveImmediate(originalWorkspace)
        
        // Simulate partial corruption (manually corrupt the workspace file)
        try store.writeCorruptWorkspace(scenario: .partiallyCorruptTab)
        
        let recovered = try store.load()
        
        // We should recover at least some tabs, never return empty when data exists
        XCTAssertNotNil(recovered, "Must not return nil on partial corruption")
        XCTAssertGreaterThan(recovered?.tabs.count ?? 0, 0, "Must salvage at least one valid tab")
        
        // The key lesson: NEVER silently discard all user data
        // Even if only 1 of 3 tabs can be parsed, we preserve that 1 tab
    }
    
    // MARK: - Atomic Write with Backup
    
    func testAtomicWrite() async throws {
        let workspace1 = WorkspaceStore.Workspace(
            tabs: [WorkspaceStore.Tab(name: "Tab 1", panes: [], layoutPreset: "single")],
            selectedTabIndex: 0
        )
        
        let workspace2 = WorkspaceStore.Workspace(
            tabs: [WorkspaceStore.Tab(name: "Tab 2", panes: [], layoutPreset: "single")],
            selectedTabIndex: 0
        )
        
        try await store.saveImmediate(workspace1)
        try await store.saveImmediate(workspace2)
        
        // Verify backup exists
        let backupURL = tempDirectory
            .appendingPathComponent("workspace.json.backup")
        
        XCTAssertTrue(FileManager.default.fileExists(atPath: backupURL.path), "Backup should exist")
        
        // Backup should contain the first workspace
        let backupData = try Data(contentsOf: backupURL)
        let backup = try JSONDecoder().decode(WorkspaceStore.Workspace.self, from: backupData)
        
        XCTAssertEqual(backup.tabs.first?.name, "Tab 1", "Backup should contain previous workspace")
    }
    
    // MARK: - Acceptance Criteria
    
    func testAcceptanceCriteria_RestoreFixtures() async throws {
        // M4 Acceptance: "contracts/fixtures/restore/ tests green"
        
        // Test multiple restore scenarios
        let scenarios = [
            WorkspaceStore.Workspace(
                tabs: [WorkspaceStore.Tab(name: "Single", panes: [
                    WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/")
                ], layoutPreset: "single")],
                selectedTabIndex: 0
            ),
            WorkspaceStore.Workspace(
                tabs: [
                    WorkspaceStore.Tab(name: "Tab 1", panes: [
                        WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/tmp")
                    ], layoutPreset: "single"),
                    WorkspaceStore.Tab(name: "Tab 2", panes: [
                        WorkspaceStore.Pane(shell: "/bin/bash", cwd: "/usr"),
                        WorkspaceStore.Pane(shell: "/bin/zsh", cwd: "/var")
                    ], layoutPreset: "columns")
                ],
                selectedTabIndex: 1
            )
        ]
        
        for scenario in scenarios {
            try await store.saveImmediate(scenario)
            let loaded = try store.load()
            
            XCTAssertEqual(loaded, scenario, "Restore fixture must match exactly")
        }
    }
}

File: windows/YOLOTerm.Tests/Persistence/WorkspaceStoreTests.cs
using System;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Xunit;
using YOLOTerm.Core.Persistence;

namespace YOLOTerm.Tests.Persistence;

public class WorkspaceStoreTests : IDisposable
{
    private readonly string testDir;
    private readonly WorkspaceStore store;
    
    public WorkspaceStoreTests()
    {
        testDir = Path.Combine(Path.GetTempPath(), $"yoloterm-test-{Guid.NewGuid()}");
        Directory.CreateDirectory(testDir);
        store = new WorkspaceStore(testDir);
    }
    
    public void Dispose()
    {
        if (Directory.Exists(testDir))
        {
            Directory.Delete(testDir, recursive: true);
        }
    }
    
    [Fact]
    public async Task Load_NoFile_ReturnsNull()
    {
        // Act
        var workspace = await store.LoadAsync();
        
        // Assert
        Assert.Null(workspace);
    }
    
    [Fact]
    public async Task Save_And_Load_RoundTrip()
    {
        // Arrange
        var workspace = new WorkspaceStore.Workspace
        {
            Tabs = new()
            {
                new WorkspaceStore.Tab(
                    Guid.NewGuid(),
                    "Tab 1",
                    new()
                    {
                        new WorkspaceStore.Pane(Guid.NewGuid(), "pwsh", @"C:\Users\Test", "Test Pane")
                    },
                    "single"
                )
            },
            SelectedTabIndex = 0
        };
        
        // Act
        await store.SaveImmediateAsync(workspace);
        var loaded = await store.LoadAsync();
        
        // Assert
        Assert.NotNull(loaded);
        Assert.Single(loaded.Tabs);
        Assert.Equal("Tab 1", loaded.Tabs[0].Name);
        Assert.Equal("single", loaded.Tabs[0].LayoutPreset);
        Assert.Single(loaded.Tabs[0].Panes);
        Assert.Equal("pwsh", loaded.Tabs[0].Panes[0].Shell);
        Assert.Equal(@"C:\Users\Test", loaded.Tabs[0].Panes[0].Cwd);
    }
    
    [Fact]
    public async Task Save_Multiple_PreservesAllTabs()
    {
        // Arrange
        var workspace = new WorkspaceStore.Workspace
        {
            Tabs = new()
            {
                new WorkspaceStore.Tab(Guid.NewGuid(), "Tab 1", new(), "auto"),
                new WorkspaceStore.Tab(Guid.NewGuid(), "Tab 2", new(), "columns"),
                new WorkspaceStore.Tab(Guid.NewGuid(), "Tab 3", new(), "rows")
            },
            SelectedTabIndex = 1
        };
        
        // Act
        await store.SaveImmediateAsync(workspace);
        var loaded = await store.LoadAsync();
        
        // Assert
        Assert.NotNull(loaded);
        Assert.Equal(3, loaded.Tabs.Count);
        Assert.Equal(1, loaded.SelectedTabIndex);
    }
    
    [Fact]
    public async Task GracefulDegradation_InvalidJson_ReturnsEmptyWorkspace()
    {
        // Arrange
        await store.WriteCorruptWorkspaceAsync(WorkspaceStore.CorruptWorkspaceScenario.InvalidJson);
        
        // Act
        var loaded = await store.LoadAsync();
        
        // Assert
        Assert.NotNull(loaded);
        Assert.Empty(loaded.Tabs);
    }
    
    [Fact]
    public async Task GracefulDegradation_MissingTabs_ReturnsEmptyWorkspace()
    {
        // Arrange
        await store.WriteCorruptWorkspaceAsync(WorkspaceStore.CorruptWorkspaceScenario.MissingTabs);
        
        // Act
        var loaded = await store.LoadAsync();
        
        // Assert
        Assert.NotNull(loaded);
        Assert.Empty(loaded.Tabs);
    }
    
    [Fact]
    public async Task GracefulDegradation_PartiallyCorruptTab_RecoverValidTabs()
    {
        // Arrange
        await store.WriteCorruptWorkspaceAsync(WorkspaceStore.CorruptWorkspaceScenario.PartiallyCorruptTab);
        
        // Act
        var loaded = await store.LoadAsync();
        
        // Assert - Should recover the valid tab, skip the corrupt one
        Assert.NotNull(loaded);
        Assert.Single(loaded.Tabs); // Only one valid tab
        Assert.Equal("Tab 1", loaded.Tabs[0].Name);
    }
}
